### ISSUE TITLE
Numba version of mc_sample_path

### ISSUE THREAD

**Issue by oyamad** [QuantEcon/QuantEcon.py#137]
Hi all,
I played further with Numba, by numba-fying? [`mc_sample_path`](https://github.com/QuantEcon/QuantEcon.py/blob/master/quantecon/mc_tools.py#L279).
The code and performance tests are found [here](http://nbviewer.ipython.org/github/oyamad/mc_sample_path_numba/blob/master/mc_sample_path_numba02.ipynb).

The Numba version can be more than 100x faster than the original `qe.mc_sample_path` depending on the problem size (and the environment).

0. Cell [3] contains the Numba version `mc_sample_path` and the NumPy version `mc_sample_path_numpy`, which follows implementation approach 1 in [@albop 's comment](https://github.com/QuantEcon/QuantEcon.py/issues/132#issuecomment-90955976).

0. `mc_sample_path_2` in cell [15] follows the approach of Code 1 in [my comment](https://github.com/QuantEcon/QuantEcon.py/issues/132#issuecomment-90792658).

As always, there is a tradeoff in introducing Numba: speed v.s. clarity/transparency. Would the use of Numba be justified in this case?

Some remarks:

* The main bottleneck in the original `qe.mc_sample_path`, compared to the NumPy version, is in random sampling, which is done for `sample_size` times. The NumPy version is faster as it calls a random sampling routine only once, with size= sample_size.

* It is interesting that the speed of the original and the NumPy versions, both of which rely on [`np.searchsorted`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.searchsorted.html), is almost constant in the size of the matrix.

**Comment by jstac** [comment]
@oyamad This is great.  You write beautiful code.  Is that a bisection search in the inner loop of the `mc_sample_path` routine?

I don't have a very strong opinion over options 1 and 2 above (cell [3] vs cell [15]).  I guess I like [15] a little better because everything to do with numba is wrapped up in a single block of code where the first line makes it immediately clear this only concerns numba optimization.

I think the use of numba would most certainly be justified in this case.  Anything over an order of magnitude is worth looking at, and anything that gets two orders of magnitude should certainly be implemented.

My only concern is that people will have older versions of numba, where generation of random variables inside the function is slow...

**Comment by mmcky** [comment]
@oyamad This is indeed a great example to consider regarding the style discussions we have been having in the #132 discussions. Thanks for adding this to the discussion. Here are my thoughts ...

I think box ``[3]`` demonstrates @albop style suggestion really well given numba ``0.17`` supports the ``np.random.random()`` line. I think ``[3]``  is pretty clear - two functions with the same interfaces doing the same job. One addition is that we should probably add a docstring flag requirement in the ``mc_sample_path`` to ensure clear identification as a numba routine. 

However it does have code duplication.

Comments on ``[15]``:
  1. Unless ``_mc_sample_path_jit()`` is useful for other functions I would vote for it to be a sub-function OR updated with a different name as the function interfaces are different between ``mc_sample_path_2`` and ``_mc_sample_path_jit()``. I would suggest something like ``_mc_generate_sample_path_jit()``. 
  1. For a global (module wide) function the ``_mc_sample_path_jit(cdfs, u, out)`` also has an unusual interface as it  doesn't return anything (attaching results to a passed in variable reference). I can see why this has been done (due to the setup of X) ... but I think with this type of interface makes it more suited as a sub-function within ``mc_sample_path_2``.
  1. The current ``[15]`` has the testing complication (comparing numpy version with numba version) brought up by @albop, but this could be alleviated by integrating a ``use_numba=True`` flag in the function specification with a ``if numba_installed and use_numba:`` block. 

In the end of the day I actually think it is perhaps best to allow both styles / approaches to optimizing functions in our style guide given the varied conditions under which numba will work and evolve. 

@jstac  I think coding to the latest version (within the release channel) of anaconda should be OK to make use of the latest feature set available, as it is easily fixed with ``conda update anaconda``. We could add a numba version warning if we make use of a feature contained in the latest version of numba. ``if numba_version < 0.17`` issue warning to upgrade anaconda.

**Comment by jstac** [comment]
@mmcky Thanks for those comments. Perhaps [3] is going to be simpler.  It's very nice code, very clear.  Shall we just go with that?  I don't see it's worth debating it for too long.

@oyamad Are you willing to turn this into a PR based around [3]?  I'm keen to get it merged into master, perhaps with a warning about the numba version as suggested by @mmcky

**Comment by oyamad** [comment]
@jstac @mmcky Thanks for your comments!

@jstac It is just standard binary search. np.searchsorted seems to implement it, as in [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/item_selection.c#L1599) and [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/npysort/binsearch.c.src#L38), although the C code there is hard (for me) to read.

@mmcky Putting the jitted function `_mc_sample_path_jit` within `mc_sample_path` as a subfunction seems not to contribute any speedup. I don't know why (maybe our Numba expert @albop could make a comment on this?).

I guess that np.random.random() in the code, which is outside the loop, is unrelated to the "support of np.random routines" in recent versions of Numba; it may be that recent versions are aware of types of returned arrays or any arrays created in intermediate steps in those routines that are used in for loops, whereas code outside loops is run in object mode anyway (again I am not sure).

It would be nice if both styles/approaches are allowed, since there are constraints in automatic [loop-jitting](http://numba.pydata.org/numba-doc/0.18.2/glossary.html#term-loop-jitting), one of which is binding in the case of `gth_solve`, which I will explain some day soon when I update the PR, although I totally agree there would be testing issues we have to take care of.

I will make a PR with [3], perhaps tomorrow.

**Comment by jstac** [comment]
Thanks @oyamad 

I second allowing both styles assuming testing issues are fixed (which is indeed important, since it's such a natural way to test the code).  I'm looking forward to seeing a PR based on either one.

**Comment by mmcky** [comment]
@oyamad Hey thanks for letting me know re: sub function ... I was able to replicate what you are talking about - I will take a closer look into this.

**Comment by oyamad** [comment]
Let me modify my previous claim: if one puts the jitted function inside `mc_sample_path` as a sub-function, there is some speedup, but not as much as in the previous case with a global function; see [here](http://nbviewer.ipython.org/github/oyamad/mc_sample_path_numba/blob/master/sub_function02.ipynb) for `mc_sample_path_3` defined in [4] whose performance is indicated on the 4th line in each `timeit` test.

**Comment by albop** [comment]
Sorry for the slow reaction. I am starting to look at the code and have one obvious question. Is the reason why you want to speed up the code, because you are simulating over a very long period ? Or is it because, you want to run many simulations ? (like 100000)
In the latter case, there may be a payoff in vectorizing over many simulations, ni which case numpy may do a good job. Or at least, it is worth checking that the whole function is jitted in nopython model. It is currently not the case since cumsum is not understood by numba.

**Comment by albop** [comment]
@oyamad : the numbafied code seems close enough to the optimal. 
I tried to compare it with a cython version, which is supposed to be execute as fast as C code:

```cython
%%cython --compile-args=-O3  --link-args=-O3
cimport cython

import numpy as np

@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
def mc_sample_path_cython(double [:,:] P, int init, int sample_size):
    # CDFs, one for each row of P
    cdef double [:,:] cdfs = np.cumsum(P, axis=-1)
    
    # Random values, uniformly sampled from [0, 1)
    cdef double [:] u = np.random.random(size=sample_size)
    
    # === set up array to store output === #
    cdef long [:] X = np.empty(sample_size, dtype=int)
    
    X[0] = init

    # === generate the sample path === #
    cdef int n = len(cdfs)
    cdef int t
    cdef int lo, hi
    
    for t in range(sample_size-1):
        lo = -1
        hi = n - 1
        while(lo < hi-1):
            m = (lo + hi) // 2
            if u[t+1] < cdfs[X[t], m]:
                hi = m
            else:
                lo = m
        X[t+1] = hi

    return X
```

The output of:

```python
sample_size = 10**5 * 2
init = 0
for i, Q in enumerate(rand_matrices):
    print 'rand_matrices[{0}] ({1} x {2})'.format(i, Q.shape[0], Q.shape[1])
#     for func in [qe.mc_sample_path, mc_sample_path_numpy, mc_sample_path, mc_sample_path_3]:
    for func in [mc_sample_path, mc_sample_path_cython]:
        %timeit func(Q, init, sample_size)
```
is: 
```
rand_matrices[0] (10 x 10)
100 loops, best of 3: 7.74 ms per loop
100 loops, best of 3: 7.18 ms per loop
rand_matrices[1] (100 x 100)
100 loops, best of 3: 14.6 ms per loop
100 loops, best of 3: 14.1 ms per loop
rand_matrices[2] (1000 x 1000)
10 loops, best of 3: 37.6 ms per loop
10 loops, best of 3: 36.2 ms per loop
rand_matrices[3] (3000 x 3000)
10 loops, best of 3: 100 ms per loop
10 loops, best of 3: 95.3 ms per loop
```
And the difference is negligible. The is really outstanding since it means that LLVM is now close enough to gcc (which cython uses), and that the translated code has very low overhead.

**Comment by albop** [comment]
Latest version of numba has a tool to annotate the python code and quickly see what runs in python mode, and which loops are automatically lifted (i.e. ran in nopython mode). 
On your example one see that the gains come from the auto-lifting : http://bl.ocks.org/albop/114cfdb9a419cc493dd5

Now, I believe that it would be useful to have the function doing the actual simulaion run in nopython entirely, so that one can loop around it. Is it really useful to keep the ability to supply init, either as a point or as a distribution ? Eventually it gets converted into a scalar index, which seems easy to do outside of the function. 

From an api perspective, what about a function like:

`simulate_markov_chain(P, init, sample_size, n_exp=None)` which would simulate a markov chain characterized by matrix `P` over `sample_size` periods and:
- perform one single simulation starting at `init` if `init` is an integer and `n_exp` is `None`
- simulate many draws if `init` is a vector (`len(init)` draws) or n_exp is an integer  (`n_exp` draws)

**Comment by jstac** [comment]
@albop "Is the reason why you want to speed up the code, because you are simulating over a very long period ? Or is it because, you want to run many simulations ? (like 100000)"

This routine is orientated to the former.  We could also add one for the latter.  In that case vectorization would probably play a bigger role.

**Comment by mmcky** [comment]
@oyamad Happy to help get this improvement merged. When convenient would you be able to set-up a branch and push the work you've done on issues #132 and #137?  I will then work on implementing issue #133 to incorporate package wide warning. I can also assist with writing the tests etc. 

@oyamad @albop @jstac I propose we focus on these two cases (#132, #137) that @oyamad has worked on to consolidate the discussions we have had with regards to style and the best ways to put some structure around supporting numba routines. Once we merge the branch I will summarise lessons learnt and add a section in the style guide on: http://quantecon.org/.

**Comment by jstac** [comment]
Sounds great.  Looking forward to seeing the code.

**Comment by oyamad** [comment]
I am copying from PR #140 the issues to discuss:

0. The loop version of `np.searchsorted` for jitting (tentatively called `search_cdf`) is pulled out from the main routine, where auto loop-jitting still works (I confirmed this by the html annotation feature; thanks @albop for the information!).

0. `search_cdf` might better be called `searchsorted`, in which case, it should work like `np.searchsorted` (the current code assumes that `v` is always smaller than `cdf[-1]`). (Docsting is yet to be written.)

0. I haven't written code for switching to a NumPy version. The current `mc_sample_path` works with NumPy by only redefining `search_cdf` to be `lambda cdf, v: np.searchsorted(cdf, v, side='right')`, so I was hesitating duplicating the whole code.

0. `cdfs` once computed may be stored in the `MarkovChain` instance. In that case, the code should be placed in the `simulate` method of `MarkovChain`, and `mc_sample_path` will just return `MakovChain(P).simulate(init, sample_size)`.

0. Motivated by a comment by @albop I was thinking of changing the API, for example by giving `MarkovChain` a method `replicate` in addition to `simulate`, as in [LinearStateSpace](https://github.com/QuantEcon/QuantEcon.py/blob/master/quantecon/lss.py#L68).

0. This is a different issue, but I wonder if there would be demand for a version of this function that is specialized for a sparse matrix input. That would use `search_cdf`, too.

0. Test is still missing.

**Comment by sanguineturtle** [comment]
@jstac I have implemented #133 with the addition of ``common_messages.py`` module. Imports can now use ``from .common_messages import numba_import_fail_message`` for use in warnings etc.

@oyamad I have adjusted the ``mc_tools.py`` file to incorporate the case where numba is not installed. As part of this I have added your previous ``mc_sample_path_numpy`` function with logic around numba for passing the numba function through ``jit``.  

@oyamad I have also temporarily re-added the version of ``mc_sample_path`` without the separate ``search_cdf`` and it is called ``mc_sample_path_jit`` [ It is also surrounded by `` ## --> ``` comments]. On my machine I find that this version is much faster than the version that is split with a separate call to search_cdf. Do you observe the same? (http://nbviewer.ipython.org/gist/sanguineturtle/8573504e4521909c0eec)

**Comment by oyamad** [comment]
Thanks @mmcky

What is the version of Numba on your machine?

**Comment by sanguineturtle** [comment]
@oyamad Sorry - I should have put that in my gist. 

```ipython
In [1]: import numba

In [2]: numba.__version__
Out[2]: '0.17.0'
```

**Comment by jstac** [comment]
@mmcky / @sanguineturtle Great, thanks, that looks good.

**Comment by mmcky** [comment]
@jstac What do you think about setting up a simple ``external.py`` module to import numba once for the whole package - rather than having this at the top of every file.

```python
numba_installed = True
try:
    from numba import jit
except ImportError:
    numba_installed = False
    jit = None
    from .common_messages import numba_import_fail_message
    warnings.warn(numba_import_fail_message, UserWarning)
```

To use it in any given module we would then use ``from .external import numba_installed, jit``. If ``numba_installed = False`` then ``jit`` would simple be ``None``

**Comment by jstac** [comment]
@mmcky I did wonder about that and I like the way you've done it.  At the same time it's another thing for developers to remember.  It seems like something people are liable to forget about down the track.  Since you ask I guess my own feeling would be to stick with what we have.

But I'm happy for you to make the call.

**Comment by oyamad** [comment]
@mmcky I reproduced your result with 0.17.0, while the both versions have almost the same speed with the latest version 0.18.2.

* [Numba 0.17.0](http://nbviewer.ipython.org/gist/oyamad/b1cb181338aaa220be6e/mc_numba_test_0_17_0.ipynb)
  `mc_sample_path` with `search_cdf` separated is much slower than `mc_sample_path_jit`.

* [Numba 0.18.2](http://nbviewer.ipython.org/gist/oyamad/b1cb181338aaa220be6e/mc_numba_test_0_18_2.ipynb)
  They exhibit almost the same speed.

**Comment by albop** [comment]
@mmcky and @oyamad: I have no idea what may cause the difference between 0.17 and 0.18.2 . Nothing strikes me in the release notes. The ability to inline functions, which should make all the difference, was added in 0.17. Maybe something subtle, possibly architecture specific.

**Comment by mmcky** [comment]
@oyamad and @albop That is interesting. I initially thought it might be due to multiple interface calls between functions - but that theory is debunked with the 0.18.2. I can't see anything in the release notes either. 

This however does raise a question ...  should we harmonise with a common quantecon development environment? I tend to use the latest ``release`` of anaconda ``2.2.0`` because I know anyone that can ``conda update anaconda`` can get the latest, and I don't have to think to much about versions. On the other hand working with the latest numba may not be a bad idea as it means the library will get faster as people upgrade in the future and it should mean less updating.

**Comment by mmcky** [comment]
@jstac Forgot to mention this when we last spoke. I think it is probably a mater of time until someone just uses ``import numba`` without the ``try`` and ``except`` frameworks anyway. I think it might be easier to have a way of calling numba within a simple module that does the work of checking if numba is importable and updating the appropriate state variables ``numba_installed`` etc. 

On the other hand it is more transparent (and less dependant) on other modules if we repeat the import logic in every file. 

Despite this I would probably lean towards ``from .external import numba_installed, jit`` as I think it is easier to remember but I think they are both intuitive enough to co-exist really. But if you have a style preference I am happy to stick to it.

**Comment by mmcky** [comment]
@oyamad @jstac @albop Given the verified improvements in ``0.18.2`` perhaps we should just stick with that and keep the functions separated (particularly if ``search_cdf`` is useful for other functions). I will delete the temporary ``mc_sample_path_jit`` function that I added tomorrow morning.

**Comment by cc7768** [comment]
@oyamad @jstac @albop I think `search_cdf` will likely be useful in other functions (I had to write my own yesterday for something else). I vote we keep it as a separate function and would be in favor of @oyamad's suggestion of changing the name to `searchsorted`.

This seems to making good progress. Thanks for getting this done

**Comment by jstac** [comment]
@oyamad @mmcky I agree with the last three comments.  As a rule l suggest we try to code to the current version of Numba in Anaconda --- but we don't have to be too strict about it.  In this case the new Numba version `0.18.2` will come soon enough and I'd rather have the nicer code made possible by the new version.

@mmcky Let's go with the suggested `from .external import numba_installed, jit` idea.

**Comment by mmcky** [comment]
Updated files to remove 0.17  ``mc_sample_path_jit()`` and changed the numba import statements to ``from .external import numba_installed, jit``. I can work on tests between ``mc_sample_path`` and ``mc_sample_path_numpy`` tomorrow.

**Comment by oyamad** [comment]
The same problem as reported [here](https://github.com/QuantEcon/QuantEcon.py/pull/103#issuecomment-95986797) occurs with the current Numba version of `mc_sample_path` with certain inputs:

```python
In [1]:

import numpy as np
import quantecon as qe
In [2]:

P_C = np.array([[0.5, 0.5], [0, 1]], order='C')
P_F = np.array([[0.5, 0.5], [0, 1]], order='F')
In [3]:

P_C
Out[3]:
array([[ 0.5,  0.5],
       [ 0. ,  1. ]])
In [4]:

P_F
Out[4]:
array([[ 0.5,  0.5],
       [ 0. ,  1. ]])
In [5]:

qe.mc_sample_path(np.array([[1.]]), init=0, sample_size=10)
Out[5]:
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
In [6]:

qe.mc_sample_path(P_C, init=1, sample_size=10)
Out[6]:
array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
In [7]:

qe.mc_sample_path(P_F, init=1, sample_size=10)
Out[7]:
array([1, 0, 1, 0, 1, 0, 1, 0, 1, 0])
```

I am going to make a fix to `mc_sample_path`.

**Comment by oyamad** [comment]
I updated [`mc_sample_path`](https://github.com/QuantEcon/QuantEcon.py/blob/0e24d870deb9c4d399f2c9ccebbb83673d596645/quantecon/mc_tools.py#L286), where memory is preallocated for `cdfs` with `order='C'` (while this is the default value, it is made explicit to highlight the issue), along with a test with matrices in C- and Fortran-contiguous orders.

**Comment by oyamad** [comment]
I changed the name `search_cdf` to `searchsorted` and made it work properly for inputs `a`, `v` such that `a[n-1] <= v`, in which case it returns `n`, where `n = len(a)`.

Being general now, this routine does not need to belong to `mc_tools.py`.

**Comment by mmcky** [comment]
@jstac @oyamad Should we move the ``searchsorted`` method to another module or sub-package? 

I would propose a ``util`` subpackage which can hold a number of modules such as ``array.py`` for array utilities. We can then promote the useful methods to the top level to match behaviour of numpy such that we can use ``qe.searchsorted()`` etc.

**Comment by mmcky** [comment]
When comparing ``mc_sample_path`` and ``mc_sample_path_numpy`` we will need to set a random seed to compare results. Has anyone come across a good way to set a package wide random seed for testing purposes, or are there better ways to deal with this issue?

**Comment by jstac** [comment]
@mmcky @oyamad I definitely support a `util` subpackage or `utilities.py` file within `quantecon`.  

I don't think setting the seed is a big deal.  Something like `np.random.seed(1234) ` should be fine.

**Comment by mmcky** [comment]
@jstac I should have documented what I had tried.  I tried using ``np.random.seed()`` but that didn't seem to work. I wasn't sure if that is because ``mc_sample_path`` uses ``np.random.random()`` - but I will look into it further. 

We can run with a ``utilities.py`` module for now. If we need to we can always refactor down the track into a sub-package.

**Comment by mmcky** [comment]
@oyamad @cc7768 ``searchsorted`` has been migrated to ``utilities.py`` and is now imported into ``mc_tools.py``. It is also currently available at the top level api ``qe.searchsorted``

**Comment by mmcky** [comment]
@oyamad I think we are close to the end of this issue. If you agree we can go ahead a submit a PR for broader review. 
  1. I have updated the docstring information to account for both ``numba`` and ``numpy`` versions, as well as the ``simulate`` method.  
  2. Updated ``test_mc_tools.py`` with a simple comparative test between ``mc_sample_path`` and ``mc_sample_path_numpy``. The issue with ``np.random.seed()`` is resolved by seeding numpy prior to calling each function so that the seed is used in the second function call.

**Comment by oyamad** [comment]
@mmcky Thanks!

I moved `test_searchsorted` from `test_mc_tools.py` to a new file `test_utilities.py`.

Just a reminder:
We also need to set up a switching for the numba function `_repeat_1d` in `cartesian.py`.
(By the way, what would be wrong if we put `@jit` on top of `cartesian` in that file?)

**Comment by mmcky** [comment]
@oyamad Good pickups  - thanks. From my review my only thought is that I think this is importing from the ``installed`` quantecon and not the local package ``quantecon``. Not a problem if we install but should we:

```python
from quantecon.utilities import searchsorted
```

be

```python
from ..utilities import searchsorted
```

I think ``nosetests`` handles relative imports. Has anyone come across a recommended style with regards to test imports? 

Thanks for pointing out ``cartesian.py``. The issue with putting ``@jit`` on the top of cartesian is that it effectively adds ``numba`` as a dependency (which it is currently in this file).  But perhaps you're wondering if that function should be accelerated? We could profile it to see where the main time is spent in that function. How about we open a new issue to track the ``cartesian`` update and get this current branch merged?


 @albop I didn't realise ``njit`` is `` jit(nopython=True)`` - good to know.

**Comment by oyamad** [comment]
0. @spencerlyon2 @mmcky Thanks for the comments. I updated `utilities.py`, where the first `searchsorted` function is now renamed to `_searchsorted`, and `_searchsorted.__doc__` is copied to `searchsorted.__doc__`.

   By the way, on IPython Notebook, `searchsorted??` does not give the code (while `searchsorted.py_func??` does). I wonder if this is a feature of Numba.

0. Regarding the test import issue, I just followed the other test files. We should discuss which style to follow.

**Comment by oyamad** [comment]
In `cartesian`, maybe list comprehension is not supported in Numba?

Anyway, I agree we should open a new issue for `cartesian`. And there has been a naming issue as discussed in #98.

**Comment by oyamad** [comment]
Sorry, I closed this by accident...

**Comment by albop** [comment]
@oyamad , @mmcky  : On the `cartesian function`, the reasoning was that the loops and list comprehensions are made over the number of dimensions, so that they are probably only a few iterations. Appart from the loop over _repeat_1d, it seems likely that everything else is going to stay in python mode, with little performance gain if any. This was just my feeling though, I didn't test it.

**Comment by albop** [comment]
As for `searchsorted`: it is a very good idea to refactor it since it is apparently used in many places. If I understand correctly, the goal is not to replicate completely numpy.searchsorted (which has several option), but provide a nopython alternative. We should keep an eye on numba progress: it should be on their roadmap.

**Comment by mmcky** [comment]
@oyamad Are you happy to issue a PR and get these improvements merged into the master branch? I think this is looking good. As mentioned above we can track any of the small remaining issues separately - it will be good to treat them outside of this branch. 

I will open new issues to track the following items
  1. testing import logic
  2. ``cartesian`` profiling (if required) and updates

**Comment by mmcky** [comment]
Hi @oyamad sorry just catching up on the weekend correspondence. I see you plan to merge ``gth_solve`` in with ``numba_improvements`` branch. Neat idea - that way we can merge all in at once. I will still open those separate issues for tracking.
